[TASK] Add Responsive Images

mneuhaus opened the pull request:
PictureViewHelper is already quite steady. Srcset is still kind of rough around the edges and only added to ```v:media.image``` so far.

xf- commented on Classes/ViewHelpers/Media/PictureViewHelper.php:
```diff
@@ -0,0 +1,92 @@
+		$this->tag->setContent($content);
+		return $this->tag->render();
+	}
+}
```
Please use new line

mneuhaus replied:
done :)

xf- commented:
Looks like edge supports src-set
http://blogs.windows.com/msedgedev/2015/06/08/introducing-srcset-responsive-images-in-microsoft-edge/

xf- commented on Classes/ViewHelpers/Resource/AbstractImageViewHelper.php:
```diff
@@ -185,4 +185,4 @@ public function preprocessSourceUri($source) {
 	}
 
-}
+}
```
newline please

mneuhaus replied:
doh! :( sorry added newline there as well

cedricziel commented:
Truly a great addition - thank you!

Can you add a test which at least executes a ViewHelper with the new argument?

Cheers,
Cedric

mneuhaus commented:
@cedricziel like this? :)

NamelessCoder commented:
Don't merge yet. Looks okay overall but has tremendous potential impact; frontend simulation is removed among other things. Needs *proper* testing as it affects every image-related ViewHelper. Some CGL issues (missing phpdoc etc) too.

mneuhaus commented:
As i said in irc, i reverted the change in the ```AbstractMediaViewHelper``` regarding ```FrontendSimulation```. 
I do use the ```FrontendSimulationUtility``` instead of the ```AbstractMediaViewHelper``` variants in the ```SourceSetViewHelperTrait``` which seems to work find in FE + BE.

+ i added missing phpdocs in ```SourceSetViewHelperTrait```. 

Any other CGL issues i'm blind about? :)

NamelessCoder commented on Classes/ViewHelpers/Media/PictureViewHelper.php:
```diff
@@ -0,0 +1,92 @@
+		$this->viewHelperVariableContainer->remove(self::SCOPE, self::SCOPE_VARIABLE_SRC);
+
+		if (FALSE === $this->viewHelperVariableContainer->exists(self::SCOPE, self::SCOPE_VARIABLE_DEFAULT_SOURCE)) {
+			throw new \Exception('Please add a source without a media query as a default.', 1438116616);
```
Should not throw a root-type `\Exception` - at the very least it should be a `\RuntimeException` since those can be caught in unit tests. In this case I suggest simply throwing a `TYPO3\CMS\Fluid\Core\ViewHelper\Exception` since that results in a friendly string output instead of a stack trace.

NamelessCoder commented on Classes/ViewHelpers/Media/PictureViewHelper.php:
```diff
@@ -0,0 +1,92 @@
+		$defaultImage->addAttribute('src', $defaultSource);
+		$defaultImage->addAttribute('alt', $this->arguments['alt']);
+
+		if ('' === $this->arguments['title']) {
```
The default value of `$this->arguments['title']` would be NULL. I would either compare `NULL === ...` or simply use `empty()`.

NamelessCoder commented on Classes/ViewHelpers/Media/SourceViewHelper.php:
```diff
@@ -0,0 +1,144 @@
+ * @subpackage ViewHelpers\Media
+ */
+class SourceViewHelper extends AbstractTagBasedViewHelper {
+	const SCOPE = 'FluidTYPO3\Vhs\ViewHelpers\Media\PictureViewHelper';
```
Line before `const`s start, please :)

NamelessCoder commented on Classes/ViewHelpers/Media/SourceViewHelper.php:
```diff
@@ -0,0 +1,144 @@
+	protected $tagName = 'source';
+
+	/**
+	 * @var \TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer
```
FQN can be imported.

NamelessCoder commented on Classes/ViewHelpers/Media/SourceViewHelper.php:
```diff
@@ -0,0 +1,144 @@
+		}
+		$result = $this->contentObject->getImgResource($src, $setup);
+
+		if ('BE' === TYPO3_MODE) {
```
I believe there is a call to restore the context after running this? If I remembered that right, then please make sure that method is called before the function `return`s.

mneuhaus replied:
could you elaborate? i don't really get what you mean, i do reset ```FrontendSimulationUtility::resetFrontendEnvironment();``` or do you mean, something that static function doesn't do?

NamelessCoder replied:
`FrontendSimulationUtility::resetFrontendEnvironment();` - yep, that's the one! Just fire that before the method returns and we're good, I think.
